develop/items/custom-data-components.md

```diff
@@ -0,0 +1,271 @@
+:::
+
+### Reading with a default value {#reading-default-value}
+
+In addition, when reading the component value we can use the `getOrDefault()` method on our `ItemStack` object to return a specified default value if the component is not present on the stack. This will safeguard against any errors resulting from a missing component. We can adjust our tooltip code like so:
+
+```java
+int clickCount = stack.getOrDefault(ModComponents.CLICK_COUNT_COMPONENT, 0);
+```
+
+As you can see this method takes two arguments, our component type like before and a default value to return if the component is not present.
+
+### Checking if a component exists {#checking-if-component-exists}
+
+You can also check for the existance of a specific component on an `ItemStack` using the `contains()` method. This takes the component type as an argument and returns `true` or `false` depending on whether the stack contains that component.
+
+```java
+boolean exists = stack.contains(ModComponents.CLICK_COUNT_COMPONENT);
+```
+
+### Fixing the error {#fixing-the-error}
+
+Let's combine those two solutions. So we will read the value with a default value of `0`, but we'll also hide the tooltip if the component is not present on the stack.
+
+@[code transcludeWith=::3](@/reference/latest/src/main/java/com/example/docs/item/custom/CounterItem.java)
+
+
+If you implement both of these solutions and hover over the item without the component you should see that it displays "Used 0 times" and no longer crashes the game.
+
+![Tooltip showing "Used 0 times"](/assets/develop/items/custom_component_2.png)
+
+Try giving yourself a Counter with our custom component removed. You can use this command to do so:
+
+```
+/give @p fabric-docs-reference:counter[!fabric-docs-reference:click_count]
+```
+
+When hovering over this item the tooltip should be missing.
+
+![Counter item with no tooltip](/assets/develop/items/custom_component_7.png)
+
+## Updating component value {#setting-component-value}
+
+Now let's try updating our component value. We're going to increase the click count each time we use our Counter item. To change the value of a component on an `ItemStack` we use the `set()` method like so:
+
+```java
+stack.set(ModComponents.CLICK_COUNT_COMPONENT, newValue);
+```
+
+This takes our component type and the value we want to set it to. In this case it will be our new click count.
+
+Let's set up a new `use()` method to read the old click count, increase it by one and then set the updated click count.
+
+@[code transcludeWith=::2](@/reference/latest/src/main/java/com/example/docs/item/custom/CounterItem.java)
+
+Now try starting the game and right clicking with the Counter. Count how many times you right click it. If you open up your inventory and look at the item again you should see that the usage number has gone up by the amount you clicked it.
```

Comment:
A few things to point out here:
1) A consistency issue: "right click" vs "right-click"
2) A grammatical issue: "you clicked it". Since the action in the past (clicking the item) has a result in the present (the counter now displays a higher value) we should use "you've clicked it" instead
3) "by the amount you clicked it" probably should be "by the amount of times you've clicked it"
4) Subjective: the sentences repeat something obvious. For example, "Count how many times you right click it." is obvious, given that we are developing a counter. I would remove this sentence. "If you open up your inventory and look at the item again..." should be "If you look at the item's tooltip (again)..." (the "again" also seems unnecessary, but I'm not certain on whether to keep it)